### DISCUSSION TITLE
How should one create links between spans?

### DISCUSSION THREAD

**Discussion by leidegre** [microsoft/aspire#6872] [General]
Typical situation. Request comes in, puts message on queue, worker picks up message.

I got two activities

1. One activity for initiating request
2. One activity for processing of message on queue

Originally, I propagated the `ActivityContext` of the initiating request but this created really long traces.

So, I wanted to test linking spans instead.

But, If I link the activity that is processing the message of the queue to the `ActivityContext` of the initiating request Aspire dashboard will present this as a backlink, when I look at the initiating request there will be a backlink to the process message activity which feels wrong but maybe that's exactly how it is supposed to work.

If I click a linked `SpanId` in the Aspire dashboard it will present a dialog where it is waiting for a span. This leads me to believe that I'm supposed to reverse this relationship and generate the SpanId for the activity that will be started and link to that before it has actually started. Otherwise, why would the Aspire dashboard think that the span might not exist?

I can't really find a good guidance on the directionality of span links or if this is what I should be doing or not. I also cannot find any good guidance on whether I want these different steps to use the same `TraceId` or not.

Any guidance would be appreciated.

**Comment by davidfowl** [comment]
cc @lmolkova

**Comment by lmolkova** [comment]
OTel links don't have direction. When you have span A (producer) and span B (consumer), the only way you can link them by adding a link to span A on the span B. 

It's not possible to create a span-id on the producer and propagate it to the consumer, then create a new Activity with this span id there. It's not supported by the public API and it's an intentional limitation (e.g. because you never know how many times the message will be processed).

> If I click a linked SpanId in the Aspire dashboard it will present a dialog where it is waiting for a span.

This is not expected - probably something else is not right. Please check if both spans are properly recorded and link points to the right span. It might also be the case that 
- your processor didn't get to that message yet 
- you have sampling enabled (and when you don't create consumer activity as a child of producer one, they are sampled independently resulting in partial traces)

**Reply by leidegre** [reply]
Maybe I misunderstand something, what is this https://github.com/dotnet/aspire/pull/4805#issuecomment-2220457875 @JamesNK Why does this behavior exist?

**Reply by JamesNK** [reply]
You see that dialog when the span linked to isn’t in aspire. The above reasons could be why. Spans are sent to aspire when they’re finished and the linked to span could be long running.

Or maybe the id of the linked span is just wrong because of programmer error.
